File: SOFTWARE/Arduino IDE/Linux/olimex_wifi_modem.py
#!/usr/bin/env python3
"""Linux driver for MOD-ESP32-C5 running the Olimex WiFi modem firmware."""

import time

import serial


class OlimexWiFiModem:
    def __init__(self, port="/dev/ttyS4", baudrate=115200, debug=True):
        self.ser = serial.Serial(port=port, baudrate=baudrate, timeout=0.05)
        self.debug = debug
        self.pending_clients = []
        self.line_fragment = ""

    def close(self):
        self.ser.close()

    def flush_input(self):
        self.ser.reset_input_buffer()

    def send_cmd(self, cmd):
        if self.debug:
            print(f"[TX] {cmd}")
        self.ser.write((cmd + "\n").encode("utf-8"))
        self.ser.flush()

    def normalize_line(self, line):
        known = (
            "PONG",
            "WIFI:CONNECTED",
            "IP:",
            "SERVER:STARTED",
            "SERVER:CLIENT:",
            "DATA:",
            "READY",
            "OK",
            "END",
            "ERROR:",
        )

        if self.line_fragment:
            line = self.line_fragment + line
            self.line_fragment = ""

        if line in ("WIFI:", "IP:", "DATA:", "SERVER:CLIENT:", "ERROR:"):
            self.line_fragment = line
            return ""

        if any(prefix.startswith(line) and prefix != line for prefix in known):
            self.line_fragment = line
            return ""

        return line

    def read_line(self, timeout=2.0):
        end = time.monotonic() + timeout
        data = bytearray()

        while time.monotonic() < end:
            b = self.ser.read(1)
            if not b:
                continue

            ch = b[0]
            if ch == ord("\r"):
                continue

            if ch == ord("\n"):
                line = data.decode("utf-8", errors="replace").strip()
                if line.startswith(">"):
                    line = line[1:].strip()
                if line:
                    return self.normalize_line(line)
                data.clear()
                continue

            if ch == ord(">") and not data:
                continue

            data.append(ch)

        line = data.decode("utf-8", errors="replace").strip()
        if line.startswith(">"):
            line = line[1:].strip()
        if line:
            return self.normalize_line(line)
        return line

    def read_exact(self, size, timeout=2.0):
        end = time.monotonic() + timeout
        data = bytearray()

        while len(data) < size and time.monotonic() < end:
            chunk = self.ser.read(size - len(data))
            if chunk:
                data.extend(chunk)

        if len(data) != size:
            raise TimeoutError(f"Expected {size} bytes, received {len(data)}")

        return bytes(data)

    def cache_client_line(self, line):
        if line.startswith("SERVER:CLIENT:"):
            self.pending_clients.append(int(line.split(":", 2)[2]))
            return True
        return False

    def wait_for(self, token, timeout=3.0):
        end = time.monotonic() + timeout
        recent = ""

        while time.monotonic() < end:
            line = self.read_line(timeout=0.2)
            if not line:
                continue

            if self.debug:
                print(f"[RX] {line}")

            self.cache_client_line(line)

            if line.startswith("ERROR:"):
                return False

            recent = (recent + line)[-128:]

            if token in line or token in recent:
                return True

        return False

    def begin(self):
        self.flush_input()
        self.send_cmd("ping")
        return self.wait_for("PONG", timeout=3.0)

    def connect_wifi(self, ssid, password):
        self.flush_input()
        self.send_cmd(f"connect {ssid} {password}")
        return self.wait_for("WIFI:CONNECTED", timeout=15.0)

    def ip(self):
        self.flush_input()
        self.send_cmd("ip")

        end = time.monotonic() + 3.0
        while time.monotonic() < end:
            line = self.read_line(timeout=0.5)
            if not line:
                continue

            if self.debug:
                print(f"[RX] {line}")

            self.cache_client_line(line)

            if line.startswith("IP:"):
                return line[3:]

        return ""

    def server_start(self, port=80):
        self.flush_input()
        self.send_cmd(f"server start {port}")
        return self.wait_for("SERVER:STARTED", timeout=3.0)

    def server_available(self):
        if self.pending_clients:
            return self.pending_clients.pop(0)

        line = self.read_line(timeout=0.05)
        if not line:
            return -1

        if self.debug:
            print(f"[RX] {line}")

        if line.startswith("SERVER:CLIENT:"):
            return int(line.split(":", 2)[2])

        return -1

    def server_read(self, client_id, timeout=2.0):
        self.send_cmd(f"server read {client_id}")
        end = time.monotonic() + timeout

        while time.monotonic() < end:
            line = self.read_line(timeout=0.2)
            if not line:
                continue

            if self.debug:
                print(f"[RX] {line}")

            if line in ("ERROR:BAD_ID", "ERROR:CLOSED"):
                return None

            if line.startswith("ERROR:"):
                raise RuntimeError(line)

            if self.cache_client_line(line):
                continue

            if not line.startswith("DATA:"):
                continue

            size = int(line[5:])
            payload = self.read_exact(size, timeout=timeout)

            trailer_end = time.monotonic() + 0.5
            while time.monotonic() < trailer_end:
                trailer = self.read_line(timeout=0.1)
                if trailer == "END":
                    break
                if trailer:
                    if self.debug:
                        print(f"[RX] {trailer}")
                    self.cache_client_line(trailer)

            return payload

        raise TimeoutError("Timed out waiting for DATA")

    def server_write(self, client_id, data):
        if isinstance(data, str):
            data = data.encode("utf-8")

        self.send_cmd(f"server write {client_id} {len(data)}")

        if not self.wait_for("READY", timeout=3.0):
            return False

        self.ser.write(data)
        self.ser.flush()

        return self.wait_for("OK", timeout=3.0)

    def server_close(self, client_id):
        self.send_cmd(f"server close {client_id}")
        return self.wait_for("OK", timeout=3.0)

File: SOFTWARE/Arduino IDE/OlimexWiFiModem/src/OlimexWiFiModem.cpp
#include "OlimexWiFiModem.h"

OlimexWiFiModem::OlimexWiFiModem(
  Stream &serial,
  Stream *debug
)
: _serial(serial),
  _debug(debug),
  _pendingClientCount(0) {
}

bool OlimexWiFiModem::begin(uint32_t timeout) {

  flushInput();

  sendCmd("ping");

  return waitFor("PONG", timeout);
}

void OlimexWiFiModem::flushInput() {

  while (_serial.available()) {
    _serial.read();
  }
}

void OlimexWiFiModem::sendCmd(const String &cmd) {

  if (_debug) {
    _debug->print("[TX] ");
    _debug->println(cmd);
  }

  _serial.print(cmd);
  _serial.print('\n');
}

void OlimexWiFiModem::cacheClient(int id) {

  if (id < 0) {
    return;
  }

  if (_pendingClientCount >= 8) {
    return;
  }

  _pendingClients[_pendingClientCount++] = id;
}

bool OlimexWiFiModem::cacheClientLine(const String &line) {

  if (!line.startsWith("SERVER:CLIENT:")) {
    return false;
  }

  cacheClient(line.substring(14).toInt());

  return true;
}

bool OlimexWiFiModem::waitFor(
  const String &token,
  uint32_t timeout
) {

  uint32_t start = millis();

  while (millis() - start < timeout) {

    String line = readLine(100);

    if (line.length()) {

      if (_debug) {
        _debug->print("[RX] ");
        _debug->println(line);
      }

      cacheClientLine(line);

      if (line.indexOf(token) >= 0) {
        return true;
      }
    }
  }

  return false;
}

String OlimexWiFiModem::readLine(uint32_t timeout) {

  String s;

  uint32_t start = millis();

  while (millis() - start < timeout) {

    while (_serial.available()) {

      char c = _serial.read();

      if (c == '\r')
        continue;

      if (c == '\n') {

        s.trim();

        if (s.length()) {

          return s;
        }

        s = "";

      } else {

        // ignore prompt chars
        if (c == '>') {

          if (s.length() == 0)
            continue;
        }

        s += c;
      }

      start = millis();
    }
  }

  s.trim();

  return s;
}

size_t OlimexWiFiModem::readBytes(
  uint8_t *buffer,
  size_t len,
  uint32_t timeout
) {

  size_t pos = 0;

  uint32_t start = millis();

  while (pos < len && millis() - start < timeout) {

    while (_serial.available() && pos < len) {

      buffer[pos++] = _serial.read();

      start = millis();
    }
  }

  return pos;
}

bool OlimexWiFiModem::connect(
  const String &ssid,
  const String &pass
) {

  flushInput();

  sendCmd("connect " + ssid + " " + pass);

  return waitFor("WIFI:CONNECTED", 15000);
}

bool OlimexWiFiModem::disconnect() {

  flushInput();

  sendCmd("disconnect");

  return waitFor("OK", 3000);
}

bool OlimexWiFiModem::isConnected() {

  flushInput();

  sendCmd("status");

  return waitFor("WIFI:CONNECTED", 3000);
}

String OlimexWiFiModem::ip() {

  flushInput();

  sendCmd("ip");

  uint32_t start = millis();

  while (millis() - start < 3000) {

    String line = readLine(3000);

    line.trim();

    if (_debug && line.length()) {
      _debug->print("[RX] ");
      _debug->println(line);
    }

    // remove prompt contamination
    if (line.startsWith(">")) {

      line.remove(0, 1);

      line.trim();
    }

    if (line.startsWith("IP:")) {

      return line.substring(3);
    }
  }

  return "";
}

bool OlimexWiFiModem::serverStart(uint16_t port) {

  flushInput();

  sendCmd("server start " + String(port));

  return waitFor("SERVER:STARTED", 3000);
}

int OlimexWiFiModem::serverAvailable() {

  if (_pendingClientCount > 0) {

    int id = _pendingClients[0];

    for (uint8_t i = 1; i < _pendingClientCount; i++) {
      _pendingClients[i - 1] = _pendingClients[i];
    }

    _pendingClientCount--;

    return id;
  }

  String line = readLine(10);

  if (line.startsWith("SERVER:CLIENT:")) {

    if (_debug) {
      _debug->print("[RX] ");
      _debug->println(line);
    }

    return line.substring(14).toInt();
  }

  return -1;
}

int OlimexWiFiModem::serverRead(
  int id,
  uint8_t *buffer,
  size_t maxLen,
  uint32_t timeout
) {

  sendCmd("server read " + String(id));

  uint32_t start = millis();

  while (millis() - start < timeout) {

    String line = readLine(100);

    if (!line.length()) {
      continue;
    }

    if (_debug) {
      _debug->print("[RX] ");
      _debug->println(line);
    }

    if (line.startsWith("ERROR:")) {
      return -1;
    }

    if (cacheClientLine(line)) {
      continue;
    }

    if (!line.startsWith("DATA:")) {
      continue;
    }

    size_t dataLen = line.substring(5).toInt();
    size_t copyLen = dataLen;

    if (copyLen > maxLen) {
      copyLen = maxLen;
    }

    size_t copied = readBytes(buffer, copyLen, timeout);

    if (copied != copyLen) {
      return -1;
    }

    uint8_t discard;

    for (size_t i = copyLen; i < dataLen; i++) {

      if (readBytes(&discard, 1, timeout) != 1) {
        return -1;
      }
    }

    waitFor("END", 500);

    return copyLen;
  }

  return -1;
}

bool OlimexWiFiModem::serverWrite(
  int id,
  const uint8_t *data,
  size_t len
) {

  sendCmd(
    "server write " +
    String(id) +
    " " +
    String(len)
  );

  if (!waitFor("READY", 3000)) {
    return false;
  }

  _serial.write(data, len);

  return waitFor("OK", 3000);
}

bool OlimexWiFiModem::serverWrite(
  int id,
  const String &s
) {

  return serverWrite(
    id,
    (const uint8_t*)s.c_str(),
    s.length()
  );
}

bool OlimexWiFiModem::serverClose(int id) {

  sendCmd("server close " + String(id));

  return waitFor("OK", 3000);
}
